daemon/containerd/image_prune.go

```
@@ -3,80 +3,172 @@ package containerd
+	"github.com/docker/docker/errdefs"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 )
 
+var imagesAcceptedFilters = map[string]bool{
+	"dangling": true,
+	"label":    true,
+	"label!":   true,
+	"until":    true,
+}
+
+// errPruneRunning is returned when a prune request is received while
+// one is in progress
+var errPruneRunning = errdefs.Conflict(errors.New("a prune operation is already running"))
+
 // ImagesPrune removes unused images
-// TODO: handle pruneFilters
 func (i *ImageService) ImagesPrune(ctx context.Context, pruneFilters filters.Args) (*types.ImagesPruneReport, error) {
+	if !atomic.CompareAndSwapInt32(&i.pruneRunning, 0, 1) {
+		return nil, errPruneRunning
+	}
+	defer atomic.StoreInt32(&i.pruneRunning, 0)
+
+	err := pruneFilters.Validate(imagesAcceptedFilters)
+	if err != nil {
+		return nil, err
+	}
+
+	danglingOnly := true
+	if pruneFilters.Contains("dangling") {
+		if pruneFilters.ExactMatch("dangling", "false") || pruneFilters.ExactMatch("dangling", "0") {
+			danglingOnly = false
+		} else if !pruneFilters.ExactMatch("dangling", "true") && !pruneFilters.ExactMatch("dangling", "1") {
+			return nil, fmt.Errorf("invalid dangling filter value: %q", pruneFilters.Get("dangling"))
+		}
+	}
+
+	filterFunc, err := i.setupFilters(ctx, pruneFilters)
+	if err != nil {
+		return nil, err
+	}
+
+	report := types.ImagesPruneReport{}
+	if !danglingOnly {
+		report, errs := i.pruneUnused(ctx, filterFunc)
+		if len(errs) > 0 {
+			return &report, combineErrors(errs)
+		}
+
+		return &report, nil
+	} else {
+		// In containerd dangling content is automatically deleted by the GC.
+		// So running prune with dangling=true is mostly a no-op, unless there
+		// was some action performed which didn't invoke the GC immediately.
+
+		// Trigger GC.
+		leases := i.client.LeasesService()
+		lease, err := leases.Create(ctx)
+		if err != nil {
+			return &report, nil
```

We don't want to return the error here?

---

We do :) Thanks!

---

We do, thanks!